Add support for NetBSD

iamleot opened the pull request:
This PR mostly mechanically adds `netbsd` to the list of platform supported by pebble.

NetBSD do not have `statfs(2)` and is neither POSIX so it also switches most Unix-es to use `unix.Statvfs()` (instead of `unix.Statfs()`) that is also defined as part of POSIX.

With these commits all tests seems to pass too (on NetBSD/amd64 9.99.98):

```
$ gmake test
go test -tags 'invariants'  -run . ./...
ok      github.com/cockroachdb/pebble   67.269s
ok      github.com/cockroachdb/pebble/bloom     0.060s
ok      github.com/cockroachdb/pebble/cmd/pebble        0.020s
?       github.com/cockroachdb/pebble/internal/ackseq   [no test files]
ok      github.com/cockroachdb/pebble/internal/arenaskl 0.332s
ok      github.com/cockroachdb/pebble/internal/base     0.078s
ok      github.com/cockroachdb/pebble/internal/batchskl 0.209s
?       github.com/cockroachdb/pebble/internal/bytealloc        [no test files]
ok      github.com/cockroachdb/pebble/internal/cache    3.728s
?       github.com/cockroachdb/pebble/internal/crc      [no test files]
?       github.com/cockroachdb/pebble/internal/datatest [no test files]
?       github.com/cockroachdb/pebble/internal/errorfs  [no test files]
ok      github.com/cockroachdb/pebble/internal/fastrand 0.006s [no tests to run]
?       github.com/cockroachdb/pebble/internal/humanize [no test files]
ok      github.com/cockroachdb/pebble/internal/intern   0.006s
?       github.com/cockroachdb/pebble/internal/invariants       [no test files]
ok      github.com/cockroachdb/pebble/internal/keyspan  4.026s
ok      github.com/cockroachdb/pebble/internal/lint     124.575s
ok      github.com/cockroachdb/pebble/internal/manifest 115.133s
?       github.com/cockroachdb/pebble/internal/manual   [no test files]
ok      github.com/cockroachdb/pebble/internal/metamorphic      47.035s
ok      github.com/cockroachdb/pebble/internal/metamorphic/crossversion 0.024s
ok      github.com/cockroachdb/pebble/internal/mkbench  0.588s
?       github.com/cockroachdb/pebble/internal/pacertoy/pebble  [no test files]
?       github.com/cockroachdb/pebble/internal/pacertoy/rocksdb [no test files]
?       github.com/cockroachdb/pebble/internal/private  [no test files]
ok      github.com/cockroachdb/pebble/internal/randvar  5.895s
?       github.com/cockroachdb/pebble/internal/rangedel [no test files]
ok      github.com/cockroachdb/pebble/internal/rangekey 0.412s
ok      github.com/cockroachdb/pebble/internal/rate     0.324s
ok      github.com/cockroachdb/pebble/internal/rawalloc 0.005s [no tests to run]
ok      github.com/cockroachdb/pebble/internal/testkeys 0.901s
ok      github.com/cockroachdb/pebble/objstorage        0.013s
?       github.com/cockroachdb/pebble/objstorage/shared [no test files]
ok      github.com/cockroachdb/pebble/objstorage/sharedobjcat   2.991s
?       github.com/cockroachdb/pebble/rangekey  [no test files]
ok      github.com/cockroachdb/pebble/record    22.766s
ok      github.com/cockroachdb/pebble/replay    15.466s
ok      github.com/cockroachdb/pebble/sstable   42.445s
ok      github.com/cockroachdb/pebble/tool      0.539s
ok      github.com/cockroachdb/pebble/tool/logs 0.028s
ok      github.com/cockroachdb/pebble/vfs       6.717s
ok      github.com/cockroachdb/pebble/vfs/atomicfs      0.014s
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/pebble/2354)
<!-- Reviewable:end -->

nicktrav commented on vfs/disk_usage_unix.go:
```diff
@@ -2,16 +2,16 @@
 func (defaultFS) GetDiskUsage(path string) (DiskUsage, error) {
-	stat := unix.Statfs_t{}
-	if err := unix.Statfs(path, &stat); err != nil {
+	stat := unix.Statvfs_t{}
```
Wont this break the other existing Unix builds? For example, I couldn't see `Statvfs_t` for Darwin or FreeBSD.

Suggest taking the approach that we took for Linux, and making a `disk_usage_netbsd.go` with the `netbsd` build tag, and then modifying the function we have here to call `Statvfs` with a `Statvfs_t` struct instead.

iamleot replied:
Argh, great catch (and yay CI/CD pipeline!). I've tested only on NetBSD. Definitely that sounds safer and I will adjust it in a completely different `disk_usage_netbsd.go` so we are sure we don't break possible existing platforms!

Thanks!

iamleot replied:
Yes, I can confirm it will break all other Unix builds except Solaris. I have reimplemented it as suggested.

Thank you Nick!

(Marking as resolved, both suggestions should be applied feel free to reopen it if there is anything left!)

nicktrav reviewed (commented):
Can you add the following to `.github/workflows/ci.yaml` to add a build target for NetBSD too?

```patch
diff --git a/.github/workflows/ci.yaml b/.github/workflows/ci.yaml
index 47325851..ead5d1fa 100644
--- a/.github/workflows/ci.yaml
+++ b/.github/workflows/ci.yaml
@@ -103,8 +103,8 @@ jobs:

     - run: go test -v ./...

-  freebsd:
-    name: go-freebsd
+  bsds:
+    name: go-bsds
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v2
@@ -119,6 +119,11 @@ jobs:
         GOOS: freebsd
       run: go build -v ./...

+    - name: NetBSD build
+      env:
+        GOOS: netbsd
+      run: go build -v ./...
+
```

Also note that we do not current test against any of the BSDs. We build, but delegate testing to folks that want run Pebble on a BSD. Currently, we only officially support the platforms that Cockroach supports: https://www.cockroachlabs.com/docs/stable/recommended-production-settings.html#software

iamleot commented:
Marking back to draft: I will work to adjust the PR as suggested by Nick, push with the force and then mark it back «Ready for review» once done and all tests passes.

iamleot commented:
> Can you add the following to `.github/workflows/ci.yaml` to add a build target for NetBSD too?
> 
> ```diff
> diff --git a/.github/workflows/ci.yaml b/.github/workflows/ci.yaml
> index 47325851..ead5d1fa 100644
> --- a/.github/workflows/ci.yaml
> +++ b/.github/workflows/ci.yaml
> @@ -103,8 +103,8 @@ jobs:
> 
>      - run: go test -v ./...
> 
> -  freebsd:
> -    name: go-freebsd
> +  bsds:
> +    name: go-bsds
>      runs-on: ubuntu-latest
>      steps:
>      - uses: actions/checkout@v2
> @@ -119,6 +119,11 @@ jobs:
>          GOOS: freebsd
>        run: go build -v ./...
> 
> +    - name: NetBSD build
> +      env:
> +        GOOS: netbsd
> +      run: go build -v ./...
> +
> ```
> 

Mmm, maybe possible suggestion: what about adding a separate `go-netbsd` job instead (like `go-freebsd`)? In that's case it's a bit nicer as user experience (i.e. we can just see a `go-freebsd` or `go-netbsd` check that fail as separate checks if only one of them will fail) and they also run in parallel (that's mostly a moot point maybe because e.g. `go-linux` checks will be likely always longer :)).

Please let me know if your prefer to have separate `go-freebsd` and `go-netbsd` jobs and I will adjust the last commit to do that instead! Thanks!

> Also note that we do not current test against any of the BSDs. We build, but delegate testing to folks that want run Pebble on a BSD. Currently, we only officially support the platforms that Cockroach supports: https://www.cockroachlabs.com/docs/stable/recommended-production-settings.html#software

That's fine! Thanks!

nicktrav commented on vfs/disk_usage_netbsd.go:
```diff
@@ -0,0 +1,26 @@
+// Copyright 2020 The LevelDB-Go and Pebble Authors. All rights reserved. Use
```
Let's update the copyright to this year, 2023.

iamleot replied:
Sure!

nicktrav reviewed (approved):
> Please let me know if your prefer to have separate go-freebsd and go-netbsd jobs and I will adjust the last commit to do that instead! Thanks!

I think what you have now is great. Less duplication of all the boilerplate for setting up the job, Go, etc. Let's leave it how it is. CI is all green here, as it stands 👍 

One more tiny nitpick on the copyright date. Once you amend, if you could squash all your commits into one and force push again, that would be great. Otherwise LGTM!

nicktrav commented:
Ah - and one more thing. You may need to rebase (a few other CI changes landed just now). Apologies if there are conflicts to resolve!

iamleot commented:
I have:

* adjusted the copyright as requested (bumped year to 2023)
* squashed all `vfs: ` commits into one
* rebased on latest `master`

I have not squashed the `ci: ` commit into `vfs: ` because IMO they are two separate changes (OK, `ci:` one depends on previous commit but if for any reason NetBSD CI support should be removed the `ci: ` commit can be eventually reverted indipendently) and squashing them will make the commits a bit weird and seems also inconsistent with the style currently adopted to me. If you think that's not okay and prefer to merge them anyway please let me know and I will and/or feel free to do that.

nicktrav commented:
LGTM. Thank you for your contribution!

Keen to hear how Pebble goes for you on *BSD.
